src/Voron/Data/Tables/TableValueCompressor.cs

```diff
@@ -0,0 +1,401 @@
+            using (var it = etagsTree.Iterate())
+            {
+                if (it.SeekToLast() == false)
+                    return; // empty table, nothing to train on
+
+                do
+                {
+                    long id = it.CreateReaderForCurrent().ReadLittleEndianInt64();
+                    table.DirectRead(id, out var size);
+                    if (size > 32 * 1024)
+                    {
+                        if (totalSkipped++ > 16 * 1024)
+                            return;  // we are scanning too much, no need to try this hard
+                        // we don't want to skip documents that are too big, they will compress
+                        // well on their own, and likely be *too* unique to add meaningfully to the
+                        // dictionary
+                        continue; 
+                    }
+
+                    sizes[dataIds.Count] = (UIntPtr)size;
+                    dataIds.Add(id);
+                    totalSize += size;
+                } while (dataIds.Count < 256 && it.MovePrev() && totalSize < 1024 * 1024);
+            }
+
+            if (dataIds.Count < 16)
+                return;// too few samples to measure
+
+            var tx = table._tx;
+            using var _ = tx.Allocator.Allocate(totalSize, out var buffer);
+            var cur = buffer.Ptr;
+            foreach (long id in dataIds)
+            {
+                var ptr = table.DirectRead(id, out var size);
+                Memory.Copy(cur, ptr, size);
+                cur += size;
+            }
+
+            using var __ = tx.Allocator.Allocate(
+                // the dictionary 
+                Constants.Storage.PageSize - PageHeader.SizeOf - sizeof(CompressionDictionaryInfo)
+                , out var dictionaryBuffer);
+
+            Span<byte> dictionaryBufferSpan = dictionaryBuffer.ToSpan();
+            ZstdLib.Train(new ReadOnlySpan<byte>(buffer.Ptr, totalSize), 
+                new ReadOnlySpan<UIntPtr>(sizes, 0, dataIds.Count), 
+                ref dictionaryBufferSpan);
+
+            var dictionariesTree = tx.CreateTree(TableSchema.DictionariesSlice);
+
+            var newId = (int)(dictionariesTree.State.NumberOfEntries + 1);
+
+            using var compressionDictionary = new ZstdLib.CompressionDictionary(newId, dictionaryBuffer.Ptr, dictionaryBufferSpan.Length, 3);
+
+            if (ShouldReplaceDictionary(tx, compressionDictionary) == false)
+            {
+                return;
+            }
+
+            table.CurrentCompressionDictionaryId = newId;
+            compressionDictionary.ExpectedCompressionRatio = GetCompressionRatio(CompressedBuffer.Length, RawBuffer.Length);
+
+            var rev = Bits.SwapBytes(newId);
+            using var _____ = Slice.External(tx.Allocator, (byte*)&rev, sizeof(int), out var slice);
+            using var ____ = dictionariesTree.DirectAdd(slice, sizeof(CompressionDictionaryInfo) + dictionaryBufferSpan.Length, out var dest);
```

Comment:
This is very crucial piece of code - I'd prefer to see explicitly the scopes of `usings` in this method. It's easier to read it IMO.